File: src/test/java/javabot/operations/KarmaOperationTest.java
package javabot.operations;

import com.antwerkz.sofia.Sofia;
import javabot.BaseMessagingTest;
import javabot.Messages;
import javabot.dao.ChangeDao;
import javabot.dao.ConfigDao;
import javabot.dao.KarmaDao;
import javabot.model.Config;
import javabot.model.Karma;
import javabot.model.ThrottleItem;
import javabot.model.UserFactory;
import org.mongodb.morphia.Datastore;
import org.pircbotx.User;
import org.testng.Assert;
import org.testng.annotations.Test;

import javax.inject.Inject;
import java.util.Date;

import static java.lang.String.format;

@Test(groups = {"operations"})
public class KarmaOperationTest extends BaseMessagingTest {
    @Inject
    private UserFactory userFactory;

    @Inject
    private KarmaDao karmaDao;

    @Inject
    private ChangeDao changeDao;

    @Inject
    private Datastore ds;

    @Inject
    private ConfigDao configDao;

    public void updateKarma() throws InterruptedException {
        Config config = configDao.get();
        Integer throttleThreshold = config.getThrottleThreshold();
        try {
            ds.delete(ds.createQuery(ThrottleItem.class));
            final Karma karma = karmaDao.find("testjavabot");
            int value = karma != null ? karma.getValue() : 0;
            final User bob = registerIrcUser("bob", "bob", "localhost");
            for (int i = 0; i < throttleThreshold; i++) {
                testMessageAs(bob, "~testjavabot++", Sofia.karmaOthersValue("testjavabot", ++value, bob.getNick()));
            }
            testMessageAs(bob, "~testjavabot++", Sofia.throttledUser());
            testMessageAs(bob, "~testjavabot--", Sofia.throttledUser());
            testMessageAs(bob, "~testjavabot--", Sofia.throttledUser());
            testMessageAs(bob, "~testjavabot--", Sofia.throttledUser());

            ds.delete(ds.createQuery(ThrottleItem.class));

            testMessageAs(bob, "~testjavabot++", Sofia.karmaOthersValue("testjavabot", ++value, bob.getNick()));
        } finally {
            config.setThrottleThreshold(throttleThreshold);
            configDao.save(config);
        }
    }

    public void noncontiguousNameReadKarma() {
        final String target = "foo " + new Date().getTime();
        testMessage(format("~karma %s", target), format("%s has no karma, %s", target, getTestUser()));
    }

    public void noncontiguousNameAddKarma() {
        final String target = "foo " + new Date().getTime();
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s ++", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
        Assert.assertTrue(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
        karmaDao.delete(karmaDao.find(target).getId());
    }

    public void karmaLooksLikeParam() {
        final String target = "foo " + new Date().getTime();
        testMessage(format("~%s--bar=as", target), Sofia.unhandledMessage(getTestUser().getNick()));
    }

    public void karmaLooksLikeParamShort() {
        testMessage("~--bar=as", Sofia.unhandledMessage(getTestUser().getNick()));
        testMessage("~ --bar=af", Sofia.unhandledMessage(getTestUser().getNick()));
    }

    public void noncontiguousNameAddKarmaTrailingSpace() {
        final String target = "foo " + new Date().getTime();
        final int karma = getKarma(target) + 1;
        testMessage(format("~%s ++", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
        Assert.assertTrue(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
        karmaDao.delete(karmaDao.find(target).getId());
    }

    public void noncontiguousNameAddKarmaWithComment() {
        final String target = "foo " + new Date().getTime();
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s++ hey coolio", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
        Assert.assertTrue(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
        karmaDao.delete(karmaDao.find(target).getId());
    }

    public void shortNameAddKarma() {
        final String target = "a"; // shortest possible name
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s++", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
        Assert.assertTrue(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
        karmaDao.delete(karmaDao.find(target).getId());
    }

    public void noNameAddKarma() {
        final String target = ""; // no name
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s++", target), Sofia.unhandledMessage(getTestUser().getNick()));
        Assert.assertFalse(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
    }

    public void noNameSubKarma() {
        final String target = ""; // no name
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) - 1;
        testMessage(format("~%s--", target), Sofia.unhandledMessage(getTestUser().getNick()));
        Assert.assertFalse(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
    }

    public void logNew() {
        final String target = new Date().getTime() + "";
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s++", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
        Assert.assertTrue(changeDao.findLog(Sofia.karmaChanged(getTestUser().getNick(), target, karma)));
        karmaDao.delete(karmaDao.find(target).getId());
    }

    public void logChanged() {
        final String target = "javabot";
        final int karma = getKarma(userFactory.createUser(target, target, "localhost").getNick()) + 1;
        testMessage(format("~%s++", target), Sofia.karmaOthersValue(target, karma, getTestUser().getNick()));
    }

    public void changeOwnKarma() {
        final int karma = getKarma(getTestUser().getNick());
        testMessage(format("~%s++", getTestUser()), "You can't increment your own karma.",
                    format("%s, you have a karma level of %d", getTestUser(), karma - 1));
        final int karma2 = getKarma(getTestUser().getNick());
        Assert.assertTrue(karma2 == karma - 1, "Should have lost one karma point.");
    }

    public void queryOwnKarma() {
        final User bill = registerIrcUser("bill", "bill", "localhost");
        final int karma = getKarma(bill.getNick());
        Assert.assertEquals(karma, 0);
        testMessageAs(bill, "~karma bill", Sofia.karmaOwnNone("bill"));
    }

    public void karmaChangeWithComments() {
        final String target = "L-----D";
        try {
            final int karma = getKarma(target);

            Messages messages = sendMessage(format("~~%s google java embedded nosql", target));
            Assert.assertFalse(messages.get(0).contains("has a karma level"), format("Should not have gotten a karma message: %s",
                                                                                     messages.get(0)));
        } finally {
            final Karma karma = karmaDao.find(target);
            if(karma != null) {
                karmaDao.delete(karma.getId());
            }
        }
    }

    private int getKarma(final String nick) {
        final Karma karma = karmaDao.find(nick);
        return karma != null ? karma.getValue() : 0;
    }
}

File: src/test/kotlin/javabot/operations/KarmaOperationTest.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import java.lang.String.format
import java.util.Date
import javabot.BaseTest
import javabot.dao.KarmaDao
import javabot.dao.NickServDao
import javabot.dao.util.EntityNotFoundException
import javabot.registerIrcUser
import javax.inject.Inject
import org.testng.Assert.assertEquals
import org.testng.Assert.assertFalse
import org.testng.Assert.assertTrue
import org.testng.annotations.DataProvider
import org.testng.annotations.Test

@Test(groups = arrayOf("operations"))
class KarmaOperationTest
@Inject
constructor(val nickServDao: NickServDao, val karmaDao: KarmaDao, val operation: KarmaOperation) :
    BaseTest() {

    @DataProvider
    fun karmaTestData() =
        arrayOf(
            arrayOf("%s++", 1),
            arrayOf("~%s++", 1),
            arrayOf("%s++", 1),
            arrayOf("~ %s ++", 1),
            arrayOf("%s--", -1),
            arrayOf("~%s--", -1),
            arrayOf("%s: ++", 1),
            arrayOf("~%s: ++", 1),
            arrayOf("~ %s: ++", 1),
            arrayOf("~ %s : ++", 1),
            arrayOf("%s: --", -1),
            arrayOf("~%s: --", -1)
        )

    @Test(dataProvider = "karmaTestData")
    fun karmaTests(command: String, expectedKarma: Int) {
        deleteKarma("foo")
        var message = message(command.format("foo"))
        var response = operation.handleMessage(message)
        var karma = getKarma("foo")
        assertEquals(expectedKarma, karma)
        deleteKarma("foo")
    }

    fun handleCPlusPlus1() {
        deleteKarma("C++")
        var message = message("C++++")
        var response = operation.handleMessage(message)
        println(response)
        var karma = getKarma("C++")
        assertEquals(1, karma)
        deleteKarma("C++")
    }

    fun handleCPlusPlus2() {
        deleteKarma("C++")
        var message = message("C++:++")
        var response = operation.handleMessage(message)
        println(response)
        var karma = getKarma("C++")
        assertEquals(1, karma)
        deleteKarma("C++")
    }

    fun noncontiguousNameReadKarma() {
        val target = "foo ${Date().time}"
        val response = operation.handleMessage(message("~karma ${target}"))
        assertEquals(response[0].value, format("%s has no karma, %s", target, TEST_USER))
    }

    fun noncontiguousNameAddKarma() {
        val target = "foo ${Date().time}"
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target} ++"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun botNameWithKarmaWithAddress() {
        val target = TEST_BOT_NICK
        val karma = getKarma(target) + 1
        val event = message("${TEST_BOT_NICK}: ${target}++")
        val response = operation.handleMessage(event)
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))

        bot.get().processMessage(event)
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun botNameWithKarma() {
        val target = TEST_BOT_NICK
        val karma = getKarma(target) + 1
        val event = message("${TEST_BOT_NICK}: ++")
        val response = operation.handleMessage(event)
        assertEquals(1, response.size)
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))

        bot.get().processMessage(event)
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun botNameKarma() {
        val target = "foo"
        val karma = getKarma(target) + 1
        val event = message("${TEST_BOT_NICK}: ${target}++", TEST_BOT_NICK)
        val response = operation.handleMessage(event)
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))

        bot.get().processMessage(event)
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun noncontiguousNameAddKarmaTrailingSpace() {
        val target = "foo ${Date().time}"
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target} ++"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun noncontiguousNameAddKarmaWithComment() {
        val target = "foo ${Date().time}"
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target}++ hey coolio"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun shortNameAddKarma() {
        val target = "a" // shortest possible name
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target}++"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun noNameAddKarma() {
        val target = "" // no name
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target}++"))
        assertEquals(response.size, 0)
        assertFalse(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
    }

    fun noNameSubKarma() {
        val target = "" // no name
        val karma = getKarma(target) - 1
        val response = operation.handleMessage(message("~${target}--"))
        assertEquals(response.size, 0)
        assertFalse(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
    }

    fun logNew() {
        val target = "${Date().time}"
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target}++"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
        assertTrue(
            changeDao.findLog(Sofia.karmaChanged(TEST_USER.nick, target, karma, TEST_CHANNEL.name))
        )
        deleteKarma(target)
    }

    fun logChanged() {
        val target = "javabot"
        val karma = getKarma(target) + 1
        val response = operation.handleMessage(message("~${target}++"))
        assertEquals(response[0].value, Sofia.karmaOthersValue(target, karma, TEST_USER.nick))
    }

    fun changeOwnKarma() {
        val karma = getKarma(TEST_USER.nick)
        val response = operation.handleMessage(message("~${TEST_USER}++"))
        assertEquals(response[0].value, "You can't increment your own karma.")
        assertEquals(response[1].value, "${TEST_USER}, you have a karma level of ${karma - 1}")
        val karma2 = getKarma(TEST_USER.nick)
        assertTrue(karma2 == karma - 1, "Should have lost one karma point.")
    }

    fun queryOwnKarma() {
        val bill = nickServDao.registerIrcUser("bill", "bill", "localhost")
        val karma = getKarma(bill.nick)
        assertEquals(karma, 0)
        val response = operation.handleMessage(message("~karma bill", user = bill))
        assertEquals(response[0].value, Sofia.karmaOwnNone("bill"))
    }

    fun karmaChangeWithComments() {
        val target = "L-----D"
        try {
            val response = operation.handleMessage(message("~target google java embedded nosql"))
            assertEquals(response.size, 0, "Should not have gotten a karma message")
        } finally {
            val karma = karmaDao.find(target)
            if (karma != null) {
                karmaDao.delete(karma.id)
            }
        }
    }

    private fun deleteKarma(nick: String) {
        try {
            karmaDao.delete(karmaDao.find(nick)?.id)
        } catch (_: EntityNotFoundException) {
            // it's okay if the target isn't found. This is a cleanup operation.
        }
    }

    private fun getKarma(nick: String): Int {
        val karma = karmaDao.find(nick)
        return if (karma != null) karma.value else 0
    }
}
